### PR TITLE
Lawful StreamT

### PR DESCRIPTION
**Pull request by ArturGajowy**
## Overview

Previous version of `matcher.StreamT` was unlawful in it not being associative. The new implementation is lawful, has the laws tested, and delegates stacksafety and laziness to the underlying F Monad while maintaining a minimal interface.

I've first tried to adapt and test the `StreamT` we have in `shared`, but gave up after seeing the amount of work needed to have the cats-law tests passing for it (even if it is associative - it's not stacksafe), and for it having an interface similar enough to what we need in rholang.

This is needed, because the effect stack in the matcher needs remodeling (see RHOL-1091), and the target effect stack will actually need associativity of the StreamT transformer.

### JIRA issue:
RHOL-1092

### Checklist
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### CODE REVIEW

**Comment by goral09** [comment]
nojs :clap:

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  it must "not trigger spurious evaluation of underlying stream for a lazy monad" in {
+    val stream: Stream[Int] = Stream.cons(1, ???)
+    var wrapped: StreamT[Coeval, Int] = null
+
+    noException shouldBe thrownBy {
+      stream.head
+      wrapped = StreamT.fromStream[Coeval, Int](Coeval.now(stream))
+    }
+
+    val next = wrapped.next
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala]: The code below could be extracted to a separate test case

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      // by requiring [F: Defer]
+    }
+  }
+
+}
+
+class StreamTLawsSpec extends CatsSuite {
+
+  type Safe[A] = Coeval[A]
+  type Eff[A] = WriterT[Safe, String, A]
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala]: What does `Eff` stand for? The "Eff monad"? If not, please change to `Effect`

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatcherMonadSpec.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    val possibleResults = Stream((0, 1), (0, 2))
+    val computation = NonDetFreeMapWithCost.fromStream(possibleResults)
+    val sum = computation.map { case (x, y) => x + y }.charge(Cost(1))
+    val (cost, _) = sum.runWithCost(Cost(possibleResults.size)).right.get
+    assert(cost.value == 0) //is 1, so just 1 is charged
+
+    val moreVariants = sum.flatMap(x => NonDetFreeMapWithCost.fromStream(Stream(x, 0, -x)))
+    val moreComputation = moreVariants.map(x => "Do sth with " + x).charge(Cost(1))
+    val (cost2, _) = moreComputation.runWithCost(Cost(possibleResults.size * 3)).right.get
+    assert(cost2.value == 0) //is 4, so only 2 is charged
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatcherMonadSpec.scala]: I don't understand the comment. What is 4?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala [hunk 4]
```diff
@@ -17,7 +17,8 @@ package object matcher {
   type OptionalFreeMapWithCost[A] = StateT[OptionWithCost, FreeMap, A]
   type OptionWithCost[A]          = OptionT[ErroredOrCostA, A]
 
-  type ErroredOrCostA[A] = StateT[Either[OutOfPhlogistonsError.type, ?], Cost, A]
+  type OOPE[A] = Either[OutOfPhlogistonsError.type, A]
```
Comment (lukasz-golebiewski) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: What does `OOPE` stand for?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala [hunk 5]
```diff
@@ -1,39 +1,105 @@
...
-import cats.implicits._
+import scala.util.{Left, Right}
+
+/**
+  * Shamelessly transcribed minimal version of Gabriel Gonzalez's beginner-friendly ListT
+  * https://github.com/Gabriel439/Haskell-List-Transformer-Library/blob/e9a1d19/src/List/Transformer.hs
+  *
+  * See also: http://www.haskellforall.com/2016/07/list-transformer-beginner-friendly-listt.html
+  *
+  * The monads resulting from applying this transformer are as lazy and stacksafe as the underlying F monad you pass.
```
Comment (lukasz-golebiewski) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala]: Do we have a code proof of this?

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala]: The laziness is checked by the test you've commented on. Thanks for your comment - I thought there is a stacksafety test in cats suite, but it's just for the tailRecM. I'll add a test for map/flatmap/combine

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala (hunk 4 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: OutOfPhlogistonError. If you want a rename, propose sth :)

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatcherMonadSpec.scala (hunk 3 above)
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatcherMonadSpec.scala]: 4 is the actual `cost2.value` we get when we run this tests against the current implementation.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala (hunk 1 above)
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala]: It's going to be changed once StreamT is made even more lazy according to the comment below - then it belongs to this test.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala (hunk 4 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: So what's the difference between `OOPE` a.k.a `OutOfPhlogistonError` and `OutOfPhlogistonsError`? :-)

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: I think you can very well read that difference from line 20 (-:

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala (hunk 2 above)
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/StreamTSpec.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala (hunk 4 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: My point is that there is a big semantic difference between those types, but their names merely differ by `s`

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: Disregard the `s` - I meant `OutOfPhlogistonsError`

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala [hunk 6]
```diff
@@ -1,39 +1,110 @@
...
-    StreamT(Monad[F].flatMap(value)(astream => astream.flatTraverse(a => f(a))))
-}
+  def empty[F[_]: Monad, A] = StreamT[F, A](Monad[F].pure(SNil()))
 
-object StreamT extends StreamTInstances0 {
-  def pure[F[_], A](single: A)(implicit F: Applicative[F]): StreamT[F, A] =
-    StreamT(F.pure(Stream(single)))
+  def pure[F[_]: Monad, A](a: A): StreamT[F, A] = StreamT(Monad[F].pure(SCons(a, empty)))
+
+  def liftF[A, F[_]: Monad](fa: F[A]): StreamT[F, A] =
```
Comment (JosDenmark) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala]:
```suggestion
  def liftF[F[_]: Monad, A](fa: F[A]): StreamT[F, A] = 
```

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala [hunk 7]
```diff
@@ -1,39 +1,110 @@
...
-  def flatMap[B](f: A => StreamT[F, B])(implicit F: Monad[F]): StreamT[F, B] =
-    flatMapF(a => f(a).value)
+  sealed trait Step[F[_], A]
+  case class SNil[F[_], A]()                              extends Step[F, A]
+  case class SCons[F[_], A](head: A, tail: StreamT[F, A]) extends Step[F, A]
 
-  def flatMapF[B](f: A => F[Stream[B]])(implicit F: Monad[F]): StreamT[F, B] =
-    StreamT(Monad[F].flatMap(value)(astream => astream.flatTraverse(a => f(a))))
-}
+  def empty[F[_]: Monad, A] = StreamT[F, A](Monad[F].pure(SNil()))
```
Comment (JosDenmark) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala]:
```suggestion
  def empty[F[_]: Monad, A]: StreamT[F, A] = StreamT[F, A](Monad[F].pure(SNil()))
```

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala (hunk 4 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: This isn't converging, moved to discord

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala (hunk 5 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala (hunk 6 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala (hunk 7 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala]: Done

**Pull request merged**